File: content/problems/2476-closest-nodes-queries-in-a-binary-search-tree/solutions/inorder-and-lower-bound/solution.cpp
class Solution {
    void inorder(TreeNode* node,vector<int>& sorted){
        if(!node)return;
        inorder(node->left,sorted);
        sorted.push_back(node->val);
        inorder(node->right,sorted);
    }
    public:vector<vector<int>> closestNodes(TreeNode* root,vector<int>& queries){
        vector<int> sorted;
        inorder(root,sorted);
        vector<vector<int>> answer;
        for(int query:queries){
            auto iterator=lower_bound(sorted.begin(),sorted.end(),query);
            int upper=iterator==sorted.end()?-1:*iterator;
            int lower=iterator!=sorted.end()&&*iterator==query?query:iterator==sorted.begin()?-1:*prev(iterator);
            answer.push_back({
                lower,upper
            });
        }
        return answer;
    }
};

File: content/problems/2476-closest-nodes-queries-in-a-binary-search-tree/solutions/inorder-and-lower-bound/Solution.java
class Solution {
    private void inorder(TreeNode node,List<Integer> sorted){
        if(node==null)return;
        inorder(node.left,sorted);
        sorted.add(node.val);
        inorder(node.right,sorted);
    }
    private int lowerBound(List<Integer> sorted,int target){
        int left=0,right=sorted.size();
        while(left<right){
            int middle=(left+right)>>>1;
            if(sorted.get(middle)<target)left=middle+1;
            else right=middle;
        }
        return left;
    }
    public List<List<Integer>> closestNodes(TreeNode root,List<Integer> queries){
        List<Integer> sorted=new ArrayList<>();
        inorder(root,sorted);
        List<List<Integer>> answer=new ArrayList<>();
        for(int query:queries){
            int index=lowerBound(sorted,query);
            int upper=index<sorted.size()?sorted.get(index):-1;
            int lower=index<sorted.size()&&sorted.get(index)==query?query:index>0?sorted.get(index-1):-1;
            answer.add(Arrays.asList(lower,upper));
        }
        return answer;
    }
}
